source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1120,13 +1121,13 @@ ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::connPool(
 
   // Note: to simplify this, we assume that the factory is only called in the scope of this
   // function. Otherwise, we'd need to capture a few of these variables by value.
-  Http::ConnectionPool::Instance& pool = container.pools_->getPool(hash_key, [&]() {
+  ConnPoolsContainer::ConnPools::OptPoolRef pool = container.pools_->getPool(hash_key, [&]() {
     return parent_.parent_.factory_.allocateConnPool(
         parent_.thread_local_dispatcher_, host, priority, protocol,
         have_options ? context->downstreamConnection()->socketOptions() : nullptr);
   });
-
-  return &pool;
+  RELEASE_ASSERT(pool.has_value(), "Pool allocation should never fail");
```

Comment:
I'm still not 100% clear on when `RELEASE_ASSERT` makes sense... I did it here because I was concerned there could be corner cases in the near future that aren't caught in testing, but would be better found explicitly in production rather than leading to undefined behaviour. I can change it to a normal `ASSERT` if that's wrong.

Reply:
Yes I would use `ASSERT` here as this is a programming error. But if it can't fail, why is this returning an optional at all? I guess it's because it can't happen right now given you can't configure limits? Yeah I would just ASSERT and add a comment.

Reply:
Yeah. I wanted to make it clear that it's not expected to fail right now in case somebody was working in this area while I work on the next phase. The assert will hopefully be gone in the next week or two, but you never know.